syntax = "proto3";

package mlops;

service MLService {
  rpc Health(HealthRequest) returns (HealthResponse);
  rpc GetModelClasses(GetModelClassesRequest) returns (GetModelClassesResponse);
  rpc TrainModel(TrainModelRequest) returns (TrainModelResponse);
  rpc Predict(PredictRequest) returns (PredictResponse);
  rpc RetrainModel(RetrainModelRequest) returns (RetrainModelResponse);
  rpc DeleteModel(DeleteModelRequest) returns (DeleteModelResponse);
  rpc ListModels(ListModelsRequest) returns (ListModelsResponse);
  rpc ListDatasets(ListDatasetsRequest) returns (ListDatasetsResponse);
  rpc LoadModel(LoadModelRequest) returns (LoadModelResponse);
}

message HealthRequest {}

message HealthResponse {
  string status = 1;
}

message GetModelClassesRequest {}

message GetModelClassesResponse {
  repeated string model_classes = 1;
}

message TrainModelRequest {
  string model_name = 1;
  string model_class = 2;
  string dataset_name = 3;
  string hyperparameters_json = 4;
  string target_column = 5;
}

message TrainModelResponse {
  bool success = 1;
  string message = 2;
}

message PredictRequest {
  string model_name = 1;
  repeated PredictDataPoint data = 2;
}

message PredictDataPoint {
  repeated float features = 1;
}

message PredictResponse {
  repeated float predictions = 1;
}

message RetrainModelRequest {
  string model_name = 1;
  string model_class = 2;
  string dataset_name = 3;
  string hyperparameters_json = 4;
  string target_column = 5;
}

message RetrainModelResponse {
  bool success = 1;
  string message = 2;
}

message DeleteModelRequest {
  string model_name = 1;
}

message DeleteModelResponse {
  bool success = 1;
  string message = 2;
}

message ListModelsRequest {}

message ListModelsResponse {
  repeated ModelInfo models = 1;
}

message ModelInfo {
  string name = 1;
  string id = 2;
  string created = 3;
  bool loaded = 4;
}

message ListDatasetsRequest {}

message ListDatasetsResponse {
  repeated DatasetInfo datasets = 1;
}

message DatasetInfo {
  string name = 1;
  int64 size = 2;
  string path = 3;
}

message LoadModelRequest {
  string model_name = 1;
}

message LoadModelResponse {
  bool success = 1;
  string message = 2;
}

